src/promotion/nhit/hash.c

```diff
@@ -0,0 +1,366 @@
+ *	|  |  |I |  |  |  |  |  |H |  | hash_map
+ *	+--+--++-+--+--+--+--+--++-+--+
+ *	     __|   rb_pointer    |        _______
+ *	    |           +        |        |     |
+ *	    v           v        v        |     v
+ *	+--++-+--+-----++-+--+--++-+--+--++-+--++-+
+ *	|  |  |  |  |X |  |  |  |  |  |  |  |  |  | ring_buffer
+ *	+--+--+--+---+-+--+--+--++-+--+--++-+--++-+
+ *		     ^           |        ^     |
+ *		     |           |________|     |
+ *		     |__________________________|
+ *
+ * ** IMPORTANT (kind of) **: Since we're not locking individual nhit_list_elems
+ * it is assumed that the rb_pointer won't overflow and catch up in time between
+ * removal from old collision and inserting new entry - that would mean that two
+ * threads would access the same empty slot at the same time. For now,
+ * the structure is reasonably sized for that to not occur, but don't try to
+ * scale it down dramatically or you gonna have a bad time.
+ */
+
+#define HASH_PRIME 4099
+
+struct nhit_list_elem {
+	ocf_core_id_t core_id;
+	uint64_t core_lba;
+
+	ocf_cache_line_t coll_prev;
+	ocf_cache_line_t coll_next;
+
+	env_atomic counter;
+};
+
+struct nhit_hash {
+	ocf_cache_line_t hash_entries;
+	uint64_t rb_entries;
+
+	ocf_cache_line_t *hash_map;
+	env_rwsem *hash_locks;
+
+	struct nhit_list_elem *ring_buffer;
+	uint64_t rb_pointer;
+	env_spinlock rb_pointer_lock;
+};
+
+ocf_error_t hash_init(uint64_t hash_size, nhit_hash_t *ctx)
+{
+	int result = 0;
+	struct nhit_hash *new_ctx;
+	uint64_t i;
+
+	new_ctx = env_vzalloc(sizeof(*new_ctx));
+	if (!new_ctx) {
+		result = -OCF_ERR_NO_MEM;
+		goto exit;
+	}
+
+	new_ctx->rb_entries = hash_size;
+	new_ctx->hash_entries = OCF_DIV_ROUND_UP(
+			new_ctx->rb_entries / 4,
+			HASH_PRIME) * HASH_PRIME - 1;
+
+	new_ctx->hash_map = env_vzalloc(
+			new_ctx->hash_entries * sizeof(*new_ctx->hash_map));
+	if (!new_ctx->hash_map) {
+		result = -OCF_ERR_NO_MEM;
+		goto dealloc_ctx;
+	}
+	for (i = 0; i < new_ctx->hash_entries; i++)
+		new_ctx->hash_map[i] = new_ctx->rb_entries;
+
+	new_ctx->hash_locks = env_vzalloc(
+			new_ctx->hash_entries * sizeof(*new_ctx->hash_locks));
+	if (!new_ctx->hash_locks) {
+		result = -OCF_ERR_NO_MEM;
+		goto dealloc_hash;
+	}
+
+	for (i = 0; i < new_ctx->hash_entries; i++) {
+		if (env_rwsem_init(&new_ctx->hash_locks[i])) {
+			result = -OCF_ERR_UNKNOWN;
+			goto dealloc_locks;
+		}
+	}
+
+	new_ctx->ring_buffer = env_vzalloc(
+			new_ctx->rb_entries * sizeof(*new_ctx->ring_buffer));
+	if (!new_ctx->ring_buffer) {
+		result = -OCF_ERR_NO_MEM;
+		goto dealloc_locks;
+	}
+	for (i = 0; i < new_ctx->rb_entries; i++) {
+		new_ctx->ring_buffer[i].core_id = OCF_CORE_ID_INVALID;
+		env_atomic_set(&new_ctx->ring_buffer[i].counter, 0);
+	}
+
+	env_spinlock_init(&new_ctx->rb_pointer_lock);
+	new_ctx->rb_pointer = 0;
+
+	*ctx = new_ctx;
+	return 0;
+
+dealloc_locks:
+	env_vfree(new_ctx->hash_locks);
+	for (i = 0; i < new_ctx->hash_entries; i++)
+		ENV_BUG_ON(env_rwsem_destroy(&new_ctx->hash_locks[i]));
```

Comment:
shouldn't it be done before freeing has_locks?

Reply:
Fixed, thanks